### DISCUSSION TITLE
electron app stuck

### DISCUSSION THREAD

**Discussion by ITfanheng** [clientIO/joint#1704] [Q&A]
Hi!

My app just like `Task Demo` with joint++.

When i change the input content and click the `clone`, the app will stuck.

Not just `clone`, some other button will execute stuck.

https://recordit.co/NyjG5OrlpR

Any help will appreciate!😊

**Comment by kumilingus** [comment]
Hello, this can be caused by storing data under the attributes of the element that are not plain objects. The `deepClone` method can fall into an infinite loop if two objects are mutual referenced.

For instance, instead of:
```js
const el1 = new dia.Element();
const el2 = new dia.Element();
/* ... */
el1.set('referenceElement', el2);
const refEl = el1.get('referenceElement');
```

do:
```js
const el1 = new dia.Element();
const el2 = new dia.Element();
/* ... */
el1.set('referenceElementId', el2.id);
const refEl = graph.getCell(el1.get('referenceElementId'))
```

Is it the case?

**Reply by ITfanheng** [reply]
Thanks, two element are independent. And all the property are string or number of the element.

**Reply by kumilingus** [reply]
So what is the result of `element.attributes` and `element.toJSON()`? Does the browser freeze if you call `joint.util.cloneDeep(element.attributes)`?

**Reply by ITfanheng** [reply]
<img width="651" alt="截屏2022-06-20 下午6 08 14" src="https://user-images.githubusercontent.com/75559678/174579263-ea759de7-45b8-4e59-9e7a-26a4dd54b795.png">

Sorry i don't know whether browser freeze.

Sometimes zoom to fit also stuck.
https://recordit.co/BJACg5vrwf

**Reply by kumilingus** [reply]
What is the `"view"` array?

Note that if there is a cellView stored on the element it will introduce a cycle reference.

```js
const model1 = view2.model

const model2 = view1.paper._views[view2.cid].model;

assert(model1 === model2);
```

**Reply by ITfanheng** [reply]
`"view"` is an object array.
<img width="641" alt="截屏2022-06-21 上午9 05 21" src="https://user-images.githubusercontent.com/75559678/174695422-d7ca9117-5bf5-4557-ba42-31e68aee9f50.png">

**Reply by kumilingus** [reply]
This is ok.

What exactly does it mean that the application is stuck? It's not clear to me from the video. Are there any errors in the console?

**Reply by ITfanheng** [reply]
Hi Roman,

sorry for the late reply. 

App is stuck means I can do nothing but exit the app. And there is no error in console.

And i find the problem is this.

<img width="608" alt="截屏2022-06-27 上午10 32 16" src="https://user-images.githubusercontent.com/75559678/175849131-f3b7bf52-6d89-425c-9bbb-20a30c70e3ff.png">

I have read the source code, `async` is `false` when addTo the graph. 

So I override the copy halo, and try to change the `async`, it works!!

But i don't konw the reason.

**Reply by kumilingus** [reply]
Note that when we add an element to the graph via `Halo` we use `async: false` to ensure the `elementView` is rendered synchronously and is available and fully rendered for anything that is coming next (e.g measuring the size of the elementView DOMElement).
